daemon/api_snaps_test.go

```diff
@@ -478,6 +478,60 @@ func (s *snapsSuite) TestPostSnapsNoWeirdses(c *check.C) {
 	}
 }
 
+func (s *snapsSuite) TestPostSnapsOptionsUnsupportedActionError(c *check.C) {
```

Comment:
I think we should also test explicitly here  that the `cleanup` to remove empty structs works. Something like:
```go
 diff --git a/daemon/api_snaps_test.go b/daemon/api_snaps_test.go
index f0f263d237..1320eba70f 100644
--- a/daemon/api_snaps_test.go
+++ b/daemon/api_snaps_test.go
@@ -426,6 +426,34 @@ func (s *snapsSuite) TestSnapsInfoFilterRemote(c *check.C) {
        c.Assert(rsp.Result, check.NotNil)
 }
 
+func (s *snapsSuite) TestPostSnapsRemoveEmptyStructs(c *check.C) {
+       var snapshotSaveCalled int
+       defer daemon.MockSnapshotSave(func(s *state.State, snaps, users []string,
+               options map[string]*snap.SnapshotOptions) (uint64, []string, *state.TaskSet, error) {
+               snapshotSaveCalled++
+
+               c.Check(snaps, check.HasLen, 2)
+               c.Check(snaps, check.DeepEquals, []string{"foo", "bar"})
+               c.Check(options, check.HasLen, 1)
+               c.Check(options, check.DeepEquals, map[string]*snap.SnapshotOptions{
+                       "foo": {Exclude: []string{"$SNAP_DATA/foo-path-1"}},
+               })
+               t := s.NewTask("fake-snapshot-2", "Snapshot two")
+               return 1, snaps, state.NewTaskSet(t), nil
+       })()
+
+       s.daemonWithOverlordMockAndStore()
+       buf := strings.NewReader(`{"action": "snapshot", "snaps": ["foo", "bar"],
+       "snapshot-options": {"foo": {"exclude":["$SNAP_DATA/foo-path-1"]}, "bar":{"exclude":[]}}}`)
+       req, err := http.NewRequest("POST", "/v2/snaps", buf)
+       c.Assert(err, check.IsNil)
+       req.Header.Set("Content-Type", "application/json")
+
+       rsp := s.asyncReq(c, req, nil)
+       c.Check(rsp.Status, check.Equals, 202)
+       c.Check(snapshotSaveCalled, check.Equals, 1)
+}
+
 func (s *snapsSuite) TestPostSnapsVerifyMultiSnapInstruction(c *check.C) {
        s.daemonWithOverlordMockAndStore()
 

```

Reply:
Thanks for this test! Implemented the proposed function. Named it `TestPostSnapsOptionsClean` and added 3rd test case `baz` for json ` "bar":{}` to cover all cleaning cases.